File: backend/src/routes/posts.ts
import { Router } from 'express';
import prisma from '../prisma';
import { AuthRequest, authMiddleware, optionalAuth } from '../middleware/auth';
import { AppError } from '../middleware/error';
import { upload } from '../middleware/upload';

const router = Router();

router.get('/', optionalAuth, async (req: AuthRequest, res, next) => {
  try {
    const limit = Math.min(parseInt(req.query.limit as string) || 20, 100);
    const offset = parseInt(req.query.offset as string) || 0;

    const posts = await prisma.post.findMany({
      take: limit,
      skip: offset,
      orderBy: { createdAt: 'desc' },
      include: {
        author: { select: { id: true, username: true, avatarUrl: true } },
        _count: { select: { likes: true, comments: true } },
      },
    });

    const total = await prisma.post.count();
    res.json({ posts, total, limit, offset });
  } catch (err) { next(err); }
});

router.get('/:id', optionalAuth, async (req: AuthRequest, res, next) => {
  try {
    const post = await prisma.post.findUnique({
      where: { id: req.params.id },
      include: {
        author: { select: { id: true, username: true, avatarUrl: true } },
        comments: {
          orderBy: { createdAt: 'asc' },
          include: { author: { select: { id: true, username: true, avatarUrl: true } } },
        },
        _count: { select: { likes: true } },
      },
    });
    if (!post) throw new AppError(404, 'Post not found');
    res.json(post);
  } catch (err) { next(err); }
});

router.post('/', authMiddleware, upload.single('image'), async (req: AuthRequest, res, next) => {
  try {
    const { text } = req.body;
    if (!text?.trim()) throw new AppError(400, 'Text is required');

    const imageUrl = req.file ? `/uploads/${req.file.filename}` : undefined;
    const post = await prisma.post.create({
      data: { text: text.trim(), imageUrl, authorId: req.userId! },
      include: { author: { select: { id: true, username: true, avatarUrl: true } } },
    });
    res.status(201).json(post);
  } catch (err) { next(err); }
});

router.delete('/:id', authMiddleware, async (req: AuthRequest, res, next) => {
  try {
    const post = await prisma.post.findUnique({ where: { id: req.params.id } });
    if (!post) throw new AppError(404, 'Post not found');
    if (post.authorId !== req.userId) throw new AppError(403, 'Not authorized');

    await prisma.post.delete({ where: { id: req.params.id } });
    res.json({ message: 'Post deleted' });
  } catch (err) { next(err); }
});

// Likes
router.post('/:id/like', authMiddleware, async (req: AuthRequest, res, next) => {
  try {
    const postId = req.params.id;
    const post = await prisma.post.findUnique({ where: { id: postId } });
    if (!post) throw new AppError(404, 'Post not found');

    const existing = await prisma.like.findUnique({
      where: { postId_userId: { postId, userId: req.userId! } },
    });

    if (existing) {
      await prisma.like.delete({ where: { id: existing.id } });
      res.json({ liked: false });
    } else {
      await prisma.like.create({ data: { postId, userId: req.userId! } });
      res.json({ liked: true });
    }
  } catch (err) { next(err); }
});

router.get('/:id/likes', optionalAuth, async (req: AuthRequest, res, next) => {
  try {
    const postId = req.params.id;
    const count = await prisma.like.count({ where: { postId } });
    let liked = false;
    if (req.userId) {
      const existing = await prisma.like.findUnique({
        where: { postId_userId: { postId, userId: req.userId } },
      });
      liked = !!existing;
    }
    res.json({ count, liked });
  } catch (err) { next(err); }
});

// Comments
router.get('/:id/comments', async (req, res, next) => {
  try {
    const limit = Math.min(parseInt(req.query.limit as string) || 20, 100);
    const offset = parseInt(req.query.offset as string) || 0;

    const comments = await prisma.comment.findMany({
      where: { postId: req.params.id },
      take: limit,
      skip: offset,
      orderBy: { createdAt: 'asc' },
      include: { author: { select: { id: true, username: true, avatarUrl: true } } },
    });
    const total = await prisma.comment.count({ where: { postId: req.params.id } });
    res.json({ comments, total, limit, offset });
  } catch (err) { next(err); }
});

router.post('/:id/comments', authMiddleware, async (req: AuthRequest, res, next) => {
  try {
    const { text } = req.body;
    if (!text?.trim()) throw new AppError(400, 'Text is required');

    const post = await prisma.post.findUnique({ where: { id: req.params.id } });
    if (!post) throw new AppError(404, 'Post not found');

    const comment = await prisma.comment.create({
      data: { text: text.trim(), postId: req.params.id, authorId: req.userId! },
      include: { author: { select: { id: true, username: true, avatarUrl: true } } },
    });
    res.status(201).json(comment);
  } catch (err) { next(err); }
});

export default router;

File: backend-py/routes/posts.py
import os
import uuid
import shutil
from fastapi import APIRouter, Depends, HTTPException, Request, UploadFile, File, Form
from sqlalchemy.orm import Session
from sqlalchemy import func
from database import get_db
from models import Post, Like, Comment, User
from auth import get_current_user_id, get_optional_user_id

router = APIRouter(prefix="/api/posts")

UPLOAD_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "uploads")
os.makedirs(UPLOAD_DIR, exist_ok=True)

ALLOWED_EXTENSIONS = {".jpg", ".jpeg", ".png", ".gif", ".webp"}


def _post_to_dict(post, likes_count, comments_count):
    return {
        "id": post.id,
        "text": post.text,
        "imageUrl": post.image_url,
        "createdAt": post.created_at.isoformat() + "Z" if post.created_at else None,
        "authorId": post.author_id,
        "author": {
            "id": post.author.id,
            "username": post.author.username,
            "avatarUrl": post.author.avatar_url,
        },
        "_count": {"likes": likes_count, "comments": comments_count},
    }


def _comment_to_dict(c):
    return {
        "id": c.id,
        "text": c.text,
        "createdAt": c.created_at.isoformat() + "Z" if c.created_at else None,
        "postId": c.post_id,
        "authorId": c.author_id,
        "author": {
            "id": c.author.id,
            "username": c.author.username,
            "avatarUrl": c.author.avatar_url,
        },
    }


@router.get("/")
def list_posts(request: Request, limit: int = 20, offset: int = 0, db: Session = Depends(get_db)):
    limit = min(limit, 100)
    posts = db.query(Post).order_by(Post.created_at.desc()).offset(offset).limit(limit).all()
    total = db.query(func.count(Post.id)).scalar()
    result = []
    for p in posts:
        lc = db.query(func.count(Like.id)).filter(Like.post_id == p.id).scalar()
        cc = db.query(func.count(Comment.id)).filter(Comment.post_id == p.id).scalar()
        result.append(_post_to_dict(p, lc, cc))
    return {"posts": result, "total": total, "limit": limit, "offset": offset}


@router.get("/{post_id}")
def get_post(post_id: str, db: Session = Depends(get_db)):
    post = db.query(Post).filter(Post.id == post_id).first()
    if not post:
        raise HTTPException(404, "Post not found")
    lc = db.query(func.count(Like.id)).filter(Like.post_id == post.id).scalar()
    comments = db.query(Comment).filter(Comment.post_id == post_id).order_by(Comment.created_at.asc()).all()
    return {
        "id": post.id,
        "text": post.text,
        "imageUrl": post.image_url,
        "createdAt": post.created_at.isoformat() + "Z" if post.created_at else None,
        "authorId": post.author_id,
        "author": {
            "id": post.author.id,
            "username": post.author.username,
            "avatarUrl": post.author.avatar_url,
        },
        "comments": [_comment_to_dict(c) for c in comments],
        "_count": {"likes": lc},
    }


@router.post("/", status_code=201)
async def create_post(
    text: str = Form(...),
    image: UploadFile = File(None),
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    if not text or not text.strip():
        raise HTTPException(400, "Text is required")
    image_url = None
    if image and image.filename:
        ext = os.path.splitext(image.filename)[1].lower()
        if ext not in ALLOWED_EXTENSIONS:
            raise HTTPException(400, "Only image files are allowed")
        filename = f"{uuid.uuid4()}{ext}"
        path = os.path.join(UPLOAD_DIR, filename)
        with open(path, "wb") as f:
            content = await image.read()
            f.write(content)
        image_url = f"/uploads/{filename}"

    post = Post(text=text.strip(), image_url=image_url, author_id=user_id)
    db.add(post)
    db.commit()
    db.refresh(post)
    return {
        "id": post.id,
        "text": post.text,
        "imageUrl": post.image_url,
        "createdAt": post.created_at.isoformat() + "Z" if post.created_at else None,
        "authorId": post.author_id,
        "author": {
            "id": post.author.id,
            "username": post.author.username,
            "avatarUrl": post.author.avatar_url,
        },
    }


@router.delete("/{post_id}")
def delete_post(post_id: str, db: Session = Depends(get_db), user_id: str = Depends(get_current_user_id)):
    post = db.query(Post).filter(Post.id == post_id).first()
    if not post:
        raise HTTPException(404, "Post not found")
    if post.author_id != user_id:
        raise HTTPException(403, "Not authorized")
    db.delete(post)
    db.commit()
    return {"message": "Post deleted"}


@router.post("/{post_id}/like")
def toggle_like(post_id: str, db: Session = Depends(get_db), user_id: str = Depends(get_current_user_id)):
    post = db.query(Post).filter(Post.id == post_id).first()
    if not post:
        raise HTTPException(404, "Post not found")
    existing = db.query(Like).filter(Like.post_id == post_id, Like.user_id == user_id).first()
    if existing:
        db.delete(existing)
        db.commit()
        return {"liked": False}
    like = Like(post_id=post_id, user_id=user_id)
    db.add(like)
    db.commit()
    return {"liked": True}


@router.get("/{post_id}/likes")
def get_likes(post_id: str, request: Request, db: Session = Depends(get_db)):
    count = db.query(func.count(Like.id)).filter(Like.post_id == post_id).scalar()
    liked = False
    user_id = get_optional_user_id(request)
    if user_id:
        existing = db.query(Like).filter(Like.post_id == post_id, Like.user_id == user_id).first()
        liked = existing is not None
    return {"count": count, "liked": liked}


@router.get("/{post_id}/comments")
def get_comments(post_id: str, limit: int = 20, offset: int = 0, db: Session = Depends(get_db)):
    limit = min(limit, 100)
    comments = db.query(Comment).filter(Comment.post_id == post_id).order_by(Comment.created_at.asc()).offset(offset).limit(limit).all()
    total = db.query(func.count(Comment.id)).filter(Comment.post_id == post_id).scalar()
    return {"comments": [_comment_to_dict(c) for c in comments], "total": total, "limit": limit, "offset": offset}


@router.post("/{post_id}/comments", status_code=201)
def create_comment(post_id: str, body: dict, db: Session = Depends(get_db), user_id: str = Depends(get_current_user_id)):
    text = body.get("text")
    if not text or not text.strip():
        raise HTTPException(400, "Text is required")
    post = db.query(Post).filter(Post.id == post_id).first()
    if not post:
        raise HTTPException(404, "Post not found")
    comment = Comment(text=text.strip(), post_id=post_id, author_id=user_id)
    db.add(comment)
    db.commit()
    db.refresh(comment)
    return _comment_to_dict(comment)
